rfcs/20190722-tflite-training.md

```diff
@@ -0,0 +1,427 @@
+
+## Goals & Non-Goals
+
+Goals:
+
+*   Illustrate a complete roadmap for TensorFlow Lite on-device training
+*   Describe user experience of TensorFlow Lite on-device training
+*   Design for functional on-device training of TensorFlow 2.x models with
+    TensorFlow Lite
+
+Non-goals:
+
+*   Quantized training on device
+*   Support TensorFlow Lite training with legacy model or TensorFlow 1.x
+
+## User Experience
+
+Throughout this document, **users** means developers who use TensorFlow Lite.
+
+This section explains how users can author a TensorFlow graph, convert a
+TensorFlow graph to TensorFlow Lite format, use TensorFlow Lite to run training
+and inference, and test TensorFlow Lite model correctness with the proposed
+design.
+
+Note: The example defines a single TensorFlow Lite model with multiple subgraphs
+for training & inference. However this isn't the only way -- You can also define
+separated training model & inference model.
+
+We're using a simplified example in this section:
+
+*   A model with 2 convolutional layers, 2 dense layers followed by a softmax
+    activation
+*   The whole model is trained offline
+*   The user want to retrain the 2 dense layers using personalized data on
+    device
+
+### Authoring TensorFlow graph
+
+**Defining the model using Keras Layers**
+
+The code defines the entire model using Keras API. Since we want to be retrain
+only dense layers on device, the model is defined in 2 parts (`conv_layers` and
+`dens_layer`).
+
+```python
+conv_layers = tf.keras.Sequential([
+  tf.keras.layers.Conv2D(10, kernel_size=(3, 3), activation="relu"),
+  tf.keras.layers.Conv2D(3, kernel_size=(3, 3), activation="relu"),
+  tf.keras.layers.Flatten(),
+])
+
+dense_layers = tf.keras.Sequential([
+  tf.keras.layers.Dense(50),
+  tf.keras.layers.Dense(10, activation='softmax'),
+])
+
+model = tf.keras.Sequential([conv_layers, dense_layers])
+```
+
+**Creating TensorFlow Functions for training and inference**
+
+With TensorFlow 2.0, the recommended way to use TensorFlow Lite is: Define a
+TensorFlow function for each invocable behavior in TensorFlow Lite. The
+converter will convert each TensorFlow function to a TensorFlow Lite subgraph,
+and users can choose subgraphs to invoke.
+
+The following code defines `inference` and `train` TensorFlow functions.
+
+```python
+@tf.function(input_signature=[
+    tf.TensorSpec(shape=[None, 64, 64, 3], dtype=tf.float32)])
+def inference(x):
+  return model(x)
+
+_LOSS_FN = tf.keras.losses.mean_squared_error
+_OPTIMIZER = tf.optimizers.RMSprop()
+@tf.function(input_signature=[
+    tf.TensorSpec(shape=[None, 64, 64, 3], dtype=tf.float32),
+    tf.TensorSpec(shape=[None, 10], dtype=tf.float32),
+])
+def train(x, y):
+  with tf.GradientTape() as tape:
+    prediction = model(x)
+    loss = _LOSS_FN(prediction, y)
+  gradients = tape.gradient(loss, model.trainable_variables)
+  _OPTIMIZER.apply_gradients(zip(gradients, model.trainable_variables))
```

Comment:
The training loop can be in the model or out of the model. Either should work and we want to support both. 

I can foresee that people may want to do both in different cases. When batch training in background, it might be desired to encode the loop into the model to simplify the logic. When training in a foreground app, application developers may want to have a custom loop to get finer grained control.